### PR TITLE
[SPARK-41167][SQL] Improve multi like performance by creating a balanced expression tree predicate

### PR DESCRIPTION
**Pull request by wankunde**
### What changes were proposed in this pull request?

Optimize like any and like all predicate expressions to a balanced expression tree predicate instead of left recursive trees predicate using exists API.
For example, expression `col like any ('abc%', '%xyz', "%mn%", "abc")`, before this PR will be optimize to `StartsWith(col, "abc") OR EndsWith(col, "xyz") OR Contains(col, "mn") OR EqualTo("abc")`, after this PR will be optimize to `(StartsWith(col, "abc") OR EndsWith(col, "xyz")) OR (Contains(col, "mn") OR EqualTo("abc"))`

### Why are the changes needed?

Local benchmark [`org.apache.spark.sql.execution.benchmark.LikeAnyBenchmark`](https://github.com/apache/spark/blob/514dd82cc30c28ad374411bbe8f6622485dae56f/sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/LikeAnyBenchmark.scala), before this PR:
```
[info] Multi like query:                         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] Query with multi like                              5149           5201          85          0.0       25747.3       1.0X
[info] Query with LikeAny simplification                  9100           9335         210          0.0       45498.0       0.6X
[info] Query without LikeAny simplification              15041          15078          48          0.0       75202.7       0.3X
```

after this PR:
```
[info] Multi like query:                         Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] Query with multi like                              5144           5145           2          0.0       25717.9       1.0X
[info] Query with LikeAny simplification                  5098           5099           0          0.0       25492.3       1.0X
[info] Query without LikeAny simplification              14501          14599         169          0.0       72504.4       0.4X
```

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Exists UT

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Seems like it became slower after your PR (?)

**Comment by wankunde** [comment]
> Seems like it became slower after your PR (?)

 Sorry for the mistake, I have update the benchmark result, after this PR,  `Query with LikeAny simplification` should be the same with `Query with multi like`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 1]
```diff
@@ -756,16 +771,16 @@ object LikeSimplification extends Rule[LogicalPlan] {
     } else {
       multi match {
         case l: LikeAll =>
-          val and = replacements.reduceLeft(And)
+          val and = combinePatterns(replacements, And.apply).get
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:774]: Could you reuse these code to create a balanced tree:
https://github.com/apache/spark/blob/5c43da6858721664318c3cdbcb051231b0e98175/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala#L1657-L1669

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 2]
```diff
@@ -743,6 +743,21 @@ object LikeSimplification extends Rule[LogicalPlan] {
     }
   }
 
+  def combinePatterns(
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:746]: Remove it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExprUtils.scala [hunk 3]
```diff
@@ -117,4 +117,23 @@ object ExprUtils extends QueryErrorsBase {
...
+  def reduceToExpressionTree(
+       patterns: Seq[Expression],
+       expressionCombiner: (Expression, Expression) => Expression): Expression = {
+    assert(patterns.size > 0)
+    var res = patterns
+    while (res.size > 1) {
+      res = res.sliding(2, 2).toSeq
+        .map(tup => if (tup.size == 2) expressionCombiner(tup.head, tup.last) else tup(0))
+    }
+    res.head
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExprUtils.scala:137]: seems that the original implementation is better than this one?
```scala
  def reduceToExpressionTree1(
      patterns: Seq[Expression],
      expressionCombiner: (Expression, Expression) => Expression): Expression = {
    var res = patterns
    while (res.size > 1) {
      res = res.sliding(2, 2).toSeq
        .map(tup => if (tup.size == 2) expressionCombiner(tup.head, tup.last) else tup(0))
    }
    res.head
  }

  def reduceToExpressionTree2(
      expressions: Seq[Expression],
      expressionCombiner: (Expression, Expression) => Expression): Expression = {
    def reduceToExpressionTree(low: Int, high: Int): Expression = high - low match {
      case 0 =>
        expressions(low)
      case 1 =>
        expressionCombiner(expressions(low), expressions(high))
      case x =>
        val mid = low + x / 2
        expressionCombiner(
          reduceToExpressionTree(low, mid),
          reduceToExpressionTree(mid + 1, high))
    }
    reduceToExpressionTree(0, expressions.size - 1)
  }

    Seq(10000, 100000, 1000000).foreach { N =>
      val expressions = Range(1, N).map(Literal(_))

      val benchmark = new Benchmark(s"Benchmark reduceToExpressionTree with $N elements", N, minNumIters = 10)
      benchmark.addCase("reduceToExpressionTree1") { _ =>
        reduceToExpressionTree1(expressions, Or.apply)
      }

      benchmark.addCase("reduceToExpressionTree2") { _ =>
        reduceToExpressionTree2(expressions, Or.apply)
      }

      benchmark.run()
    }
```

```
OpenJDK 64-Bit Server VM 1.8.0_352-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Benchmark reduceToExpressionTree with 1000000 elements:  Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
--------------------------------------------------------------------------------------------------------------------------------------
reduceToExpressionTree1                                           823           1012         212          1.2         822.5       1.0X
reduceToExpressionTree2                                           150            183          25          6.7         149.7       5.5X

```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/LikeSimplificationSuite.scala [hunk 4]
```diff
@@ -207,11 +207,17 @@ class LikeSimplificationSuite extends PlanTest {
...
     val optimized = Optimize.execute(originalQuery.analyze)
     val correctAnswer = testRelation
-      .where((((((StartsWith($"a", "abc") || EndsWith($"a", "xyz")) ||
-        (Length($"a") >= 6 && (StartsWith($"a", "abc") && EndsWith($"a", "def")))) ||
-        Contains($"a", "mn")) || ($"a" === "")) || ($"a" === "abc")) ||
-        ($"a" likeAny("abc\\%", "abc\\%def", "%mn\\%")))
-      .analyze
+      .where(
+        (
+          (
```
Comment (wangyum) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/LikeSimplificationSuite.scala:212]: Please fix the format.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/LikeAnyBenchmark.scala [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+ * To run this benchmark:
+ * {{{
+ *   1. without sbt: bin/spark-submit --class <this class>
+ *     --jars <spark core test jar>,<spark catalyst test jar> <spark sql test jar>
+ *   2. build/sbt "sql/Test/runMain <this class>"
+ *   3. generate result: SPARK_GENERATE_BENCHMARK_FILES=1 build/sbt "sql/Test/runMain <this class>"
+ *      Results will be written to "benchmarks/LikeAnyBenchmark-results.txt".
+ * }}}
+ */
+object LikeAnyBenchmark extends SqlBasedBenchmark {
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/LikeAnyBenchmark.scala:40]: Please remove this benchmark test.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExprUtils.scala [hunk 6]
```diff
@@ -117,4 +117,23 @@ object ExprUtils extends QueryErrorsBase {
...
+
+  /**
+   * Combine a number of boolean expressions into a balanced expression tree. These expressions are
+   * either combined by a logical [[And]] or a logical [[Or]].
+   *
+   * A balanced binary tree is created because regular left recursive trees cause considerable
+   * performance degradations and can cause stack overflows.
+   */
+  def reduceToExpressionTree(
+       patterns: Seq[Expression],
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExprUtils.scala]: `patterns` -> `expressions`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExprUtils.scala [hunk 7]
```diff
@@ -117,4 +117,29 @@ object ExprUtils extends QueryErrorsBase {
...
+   * Combine a number of boolean expressions into a balanced expression tree. These expressions are
+   * either combined by a logical [[And]] or a logical [[Or]].
+   *
+   * A balanced binary tree is created because regular left recursive trees cause considerable
+   * performance degradations and can cause stack overflows.
+   */
+  def reduceToExpressionTree(
+       expressions: Seq[Expression],
+       expressionCombiner: (Expression, Expression) => Expression): Expression = {
+    assert(expressions.size > 0)
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExprUtils.scala:131]: Remove this `assert`?

**Comment by wangyum** [comment]
@wankunde Please fix the PR title and description.

**Review by wangyum — APPROVED** [review]

**Comment by wangyum** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 8]
```diff
@@ -756,16 +756,16 @@ object LikeSimplification extends Rule[LogicalPlan] {
     } else {
       multi match {
         case l: LikeAll =>
-          val and = replacements.reduceLeft(And)
+          val and = buildBalancedPredicate(replacements, And.apply)
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:759]: buildBalancedPredicate(replacements, And.apply) -> buildBalancedPredicate(replacements, And)?

**Review by wangyum — APPROVED** [review]

**Comment by wankunde** [comment]
Retest this please

**Comment by wangyum** [comment]
Merged to master.

**Pull request closed without merging**